File: typescript/src/problems1_100/two-sum.ts
class Problem {
  twoSum(nums: number[], target: number): number[] {
    throw new Error(`'Not implemented (${nums}, ${target}).`);
  }
}

class BruteForce extends Problem {
  twoSum(nums: number[], target: number): number[] {
    for (let i = 0; i < nums.length; i++) {
      let complement = target - nums[i];
      if (nums.slice(i + 1).includes(complement)) {
        return [i, nums.indexOf(complement, i + 1)];
      }
    }
    return undefined;
  }
}

class SinglePass extends Problem {
  twoSum(nums: number[], target: number): number[] {
    // remaining to index mapping
    let seen = {};
    for (let i = 0; i < nums.length; i++) {
      let complement = target - nums[i];
      if (complement in seen) {
        return [seen[complement], i];
      }
      seen[nums[i]] = i;
    }
    return [-1, -1];
  }
}

export const SOLUTIONS = [new BruteForce(), new SinglePass()];

File: python/src/problems1_100/two_sum.py
from abc import ABC, abstractmethod
from typing import override


class Problem(ABC):
    @abstractmethod
    def two_sum(self, nums: list[int], target: int) -> list[int] | None:
        pass


class BruteForce(Problem):
    @override
    def two_sum(self, nums: list[int], target: int) -> list[int] | None:
        for i, num in enumerate(nums):
            complement: int = target - num
            if complement in nums[i + 1:]:
                return [i, nums.index(complement, i + 1)]
        return None


class SinglePass(Problem):
    @override
    def two_sum(self, nums: list[int], target: int) -> list[int] | None:
        # remaining to index mapping
        seen: dict[int, int] = {}
        for i, num in enumerate(nums):
            complement: int = target - num
            if complement in seen:
                return [seen[complement], i]
            seen[num] = i
        return [-1, -1]


SOLUTIONS: list[Problem] = [BruteForce(), SinglePass()]
